### PR TITLE
[Intel GPU] trigger tf32 no-gpu warn only when setting true

### LINKED ISSUE #149829: TF32 acceleration on top of oneDNN is available for Intel GPUs. The current Torch version does not have Intel GPU Support
The warning message 

> /opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/backends/mkldnn/__init__.py:78: UserWarning: TF32 acceleration on top of oneDNN is available for Intel GPUs. The current Torch version does not have Intel GPU Support. (Triggered internally at /var/lib/jenkins/workspace/aten/src/ATen/Context.cpp:148.)
  > torch._C._set_onednn_allow_tf32(_allow_tf32)

Has been triggering with a normal cpu installation of PyTorch from pypi, making it annoying and it is unclear what the user needs to do. Would it make sense to suppress or improve this warning? (For 2.7 as well)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @albanD @malfet

### PR DESCRIPTION
**Pull request by ZhiweiYan-96**
Fix issue #149829 

# Detail
In `torch.export` initialization stage, the context variable of `torch.backends.mkldn` would be initialized at function `_ignore_backend_decomps` in `torch/export/_trace.py`.

 It should be wrong to trigger no-gpu warning when trying to setting the value to `False` in a CPU-Only environment.  The right behavior is raising warning only when user try to turn it on if no GPU. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149926

### CODE REVIEW

**Comment by justinchuby** [comment]
Thanks! Would it be possible to cherry pick this into 2.7?

**Comment by ZhiweiYan-96** [comment]
> Thanks! Would it be possible to cherry pick this into 2.7?

@justinchuby  Sure, I would cherry-pick it when this PR finishes review.  Thank you again for pointing out this issue and help us enhance the quality.

**Review by EikanWang — APPROVED** [review]

File: aten/src/ATen/Context.cpp [hunk 1]
```diff
@@ -145,7 +145,9 @@ void Context::setAllowTF32OneDNN(bool b){
 #ifdef USE_XPU
   allow_tf32_onednn = b;
 #else
-  TORCH_WARN("TF32 acceleration on top of oneDNN is available for Intel GPUs. The current Torch version does not have Intel GPU Support.");
+  if(b){
```
Comment (guangyey) [inline aten/src/ATen/Context.cpp:148]:
```suggestion
  if (b) {
```

Comment (guangyey) [inline aten/src/ATen/Context.cpp:148]:
```suggestion
  if (b) {
```

Comment (guangyey) [reply aten/src/ATen/Context.cpp:148]: Please fix the lint.

Comment (EikanWang) [reply aten/src/ATen/Context.cpp:148]: Why is the linter issue not captured?

Comment (ZhiweiYan-96) [reply aten/src/ATen/Context.cpp:148]: > Please fix the lint.
I wrongly thought you use some git button to fix the lint issue since you resolve the comment... hah, I will fix them, thanks for reminding.

> Why is the lint issue not captured?

@EikanWang this file is not in `.lintrunner.toml` .  (You may see other `if(b)` like case in this file) . Maybe this is a choice of other maintainers...

Comment (ZhiweiYan-96) [reply aten/src/ATen/Context.cpp]: fixed in the new commit

Comment (guangyey) [reply aten/src/ATen/Context.cpp]: No. Refer to https://github.com/pytorch/pytorch/blob/a8d0c5c92818186119d4a94d98999acc3f549a7e/.lintrunner.toml#L55-L57, all .cpp file in aten/src/ATen are excluded from clangformat.

Comment (guangyey) [reply aten/src/ATen/Context.cpp]: > > Please fix the lint.
> 
> I wrongly thought you use some git button to fix the lint issue since you resolve the comment... hah, I will fix them, thanks for reminding.

Haha, just a reminder. I will break the ghstack merge label if I use the git button. So I revert my patch.

**Review by guangyey — APPROVED** [review]

**Comment by guangyey** [comment]
@malfet May I know if this hotfix is reasonable for you?

**Comment by guangyey** [comment]
> @malfet May I know if this hotfix is reasonable for you? It will suppress the unexpected warning message.

**Comment by ZhiweiYan-96** [comment]
Hi, @malfet could you please help take a look on this PR?  Thanks!

**Comment by guangyey** [comment]
@malfet @atalman This PR aims to suppress an unexpected warning message and is targeted for the release branch. May I know if you have any comments.

**Comment by ZhiweiYan-96** [comment]
hi, @malfet  @atalman Could you take a look at this PR when you have time? Appreciation for your suggestions.
This PR introduces minor and straightforward change to trigger the warning logic on tf32 setting. 
It will fix  https://github.com/pytorch/pytorch/issues/149829 and it is nice to be land in 2.7 release version. Thanks.

**Comment by malfet** [comment]
@guangyey do you know why this code is used in cpu-only code?

**Comment by ZhiweiYan-96** [comment]
> @guangyey do you know why this code is used in cpu-only code?

hi @malfet , it is because `torch.export` initializes flags of backends. Following is a simple backtrace from pdb.


```
-> return _export(
  /4T-720/conda_envs/zhiwei-int4/lib/python3.10/site-packages/torch/export/_trace.py(1072)wrapper()
-> ep = fn(*args, **kwargs)
  /4T-720/conda_envs/zhiwei-int4/lib/python3.10/site-packages/torch/export/exported_program.py(122)wrapper()
-> return fn(*args, **kwargs)
  /4T-720/conda_envs/zhiwei-int4/lib/python3.10/site-packages/torch/export/_trace.py(2111)_export()
-> ep = _export_for_training(
  /4T-720/conda_envs/zhiwei-int4/lib/python3.10/site-packages/torch/export/_trace.py(1072)wrapper()
-> ep = fn(*args, **kwargs)
  /4T-720/conda_envs/zhiwei-int4/lib/python3.10/site-packages/torch/export/exported_program.py(122)wrapper()
-> return fn(*args, **kwargs)
  /4T-720/conda_envs/zhiwei-int4/lib/python3.10/site-packages/torch/export/_trace.py(1973)_export_for_training()
-> export_artifact = export_func(
  /4T-720/conda_envs/zhiwei-int4/lib/python3.10/site-packages/torch/export/_trace.py(1916)_non_strict_export()
-> aten_export_artifact = _to_aten_func(  # type: ignore[operator]
  /4T-720/conda_envs/zhiwei-int4/lib/python3.10/site-packages/torch/export/_trace.py(1696)_export_to_aten_ir_make_fx()
-> with torch.nn.utils.stateless._reparametrize_module(
  /4T-720/conda_envs/zhiwei-int4/lib/python3.10/contextlib.py(142)__exit__()
-> next(self.gen)
  /4T-720/conda_envs/zhiwei-int4/lib/python3.10/site-packages/torch/export/_trace.py(173)_ignore_backend_decomps()
-> torch.backends.mkldnn.set_flags(*orig_mkldnn_flag)
````

**Comment by malfet** [comment]
@ZhiweiYan-96 imo better solution would have been for `torch._C._get_onednn_allow_tf32()` to return None if compiled without XPU

**Comment by ZhiweiYan-96** [comment]
@malfet Thanks for you suggestions, and it seems more reasonable than current change. I will check your advice and push commit, thanks.

**Comment by malfet** [comment]
@ZhiweiYan-96 any updates?

**Comment by ZhiweiYan-96** [comment]
hi @malfet I saw you push a  https://github.com/pytorch/pytorch/pull/150358 for closing the issue.  I miss the tracking here due to some urgent affairs.  Sincere appreciation for your help.

**Comment by ZhiweiYan-96** [comment]
The issue has been fixed in https://github.com/pytorch/pytorch/pull/150358
